### ISSUE TITLE
Undesired behaviour running out of WebGL contexts

### ISSUE THREAD

**Issue by janba** [jupyterlab/jupyterlab#5890]
Hi! 

Jupyter Lab produces a JS error when it runs out of WebGL contexts. At least this is true when the WebGL rendering is done via Plot.ly. Jupyter Notebook is more well behaved. /Andreas

**To Reproduce**

0. Make sure your have installed Plotly (using pip) and the jupyter lab extension for Plotly.
1. Create a notebook with just one cell containing
```
import plotly.offline as py
import plotly.graph_objs as go
py.init_notebook_mode(connected=False)
mesh = go.Mesh3d(x=[0,1,0],y=[0,0,1],z=[0,0,0],i=[0],j=[1],k=[2],flatshading=False,color='#333333')
py.iplot([mesh])
```
2. Run this cell repeatedly. Initially, this should render a dark triangle. At some point you run out of WebGL contexts (8 times on my computer) and at that point, I get the error message:

> Javascript Error: null is not an object (evaluating 'scene.container.parentNode.removeChild')

and no graphics output....

This could be a Plotly problem. However, the same is not true in Jupyter Notebook. Running out of WebGL contexts, Jupyter Notebook keeps on going - each time displaying the correct image.

**Expected behavior**

I would have thought it worked in Lab same as Notebook - just recycling the old contexts.

**Desktop (please complete the following information):**

- OS: Mac OS 10.14.2
- Browser: Safari 12.0.2
- JupyterLab: 0.35.4

**Additional context**
<details><summary>Command Line Output</summary>
<pre>
[D 21:58:00.505 LabApp] Searching ['/Users/andreas/Teaching/02580/02580-pysrc', '/Users/andreas/.jupyter', '/opt/local/Library/Frameworks/Python.framework/Versions/3.7/etc/jupyter', '/opt/local/etc/jupyter'] for config files
[D 21:58:00.505 LabApp] Looking for jupyter_config in /opt/local/etc/jupyter
[D 21:58:00.505 LabApp] Looking for jupyter_config in /opt/local/Library/Frameworks/Python.framework/Versions/3.7/etc/jupyter
[D 21:58:00.505 LabApp] Looking for jupyter_config in /Users/andreas/.jupyter
[D 21:58:00.505 LabApp] Looking for jupyter_config in /Users/andreas/Teaching/02580/02580-pysrc
[D 21:58:00.506 LabApp] Looking for jupyter_notebook_config in /opt/local/etc/jupyter
[D 21:58:00.506 LabApp] Looking for jupyter_notebook_config in /opt/local/Library/Frameworks/Python.framework/Versions/3.7/etc/jupyter
[D 21:58:00.506 LabApp] Looking for jupyter_notebook_config in /Users/andreas/.jupyter
[D 21:58:00.506 LabApp] Looking for jupyter_notebook_config in /Users/andreas/Teaching/02580/02580-pysrc
[D 21:58:00.708 LabApp] Paths used for configuration of jupyter_notebook_config: 
    	/opt/local/etc/jupyter/jupyter_notebook_config.json
[D 21:58:00.709 LabApp] Paths used for configuration of jupyter_notebook_config: 
    	/opt/local/Library/Frameworks/Python.framework/Versions/3.7/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
    	/opt/local/Library/Frameworks/Python.framework/Versions/3.7/etc/jupyter/jupyter_notebook_config.json
[D 21:58:00.709 LabApp] Paths used for configuration of jupyter_notebook_config: 
    	/Users/andreas/.jupyter/jupyter_notebook_config.json
[I 21:58:00.715 LabApp] JupyterLab extension loaded from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/jupyterlab
[I 21:58:00.715 LabApp] JupyterLab application directory is /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab
[W 21:58:00.716 LabApp] JupyterLab server extension not enabled, manually loading...
[I 21:58:00.720 LabApp] JupyterLab extension loaded from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/jupyterlab
[I 21:58:00.720 LabApp] JupyterLab application directory is /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab
[I 21:58:00.720 LabApp] Serving notebooks from local directory: /Users/andreas/Teaching/02580/02580-pysrc
[I 21:58:00.720 LabApp] The Jupyter Notebook is running at:
[I 21:58:00.720 LabApp] http://localhost:8888/?token=5d6f40f4ba80888de5202ce50122285fc7a496c0a2050ed5
[I 21:58:00.720 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 21:58:00.725 LabApp] 
    
    To access the notebook, open this file in a browser:
        file:///Users/andreas/Library/Jupyter/runtime/nbserver-69920-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=5d6f40f4ba80888de5202ce50122285fc7a496c0a2050ed5
[D 21:58:02.007 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.008 LabApp] Using contents: services/contents
[D 21:58:02.012 LabApp] 200 GET /lab?token=5d6f40f4ba80888de5202ce50122285fc7a496c0a2050ed5 (::1) 5.41ms
[D 21:58:02.017 LabApp] Path vendors~main.d5a4c517c33d84c6b8c4.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/static/vendors~main.d5a4c517c33d84c6b8c4.js
[D 21:58:02.019 LabApp] Path main.df6e80aa5d2245db9057.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/static/main.df6e80aa5d2245db9057.js
[D 21:58:02.019 LabApp] 304 GET /lab/static/main.df6e80aa5d2245db9057.js (::1) 0.51ms
[D 21:58:02.044 LabApp] 200 GET /lab/static/vendors~main.d5a4c517c33d84c6b8c4.js (::1) 27.72ms
[D 21:58:02.495 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.496 LabApp] Found kernel python3 in /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/kernels
[D 21:58:02.499 LabApp] 200 GET /api/kernelspecs?1547931482493 (::1) 4.19ms
[D 21:58:02.500 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.500 LabApp] 200 GET /api/terminals?1547931482494 (::1) 0.71ms
[D 21:58:02.527 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.529 LabApp] 200 GET /lab/api/settings/@jupyterlab/application-extension:sidebar?1547931482525 (::1) 2.13ms
[D 21:58:02.530 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.546 LabApp] 200 GET /lab/api/settings/@jupyterlab/shortcuts-extension:plugin?1547931482526 (::1) 15.93ms
[D 21:58:02.546 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.548 LabApp] 200 GET /lab/api/settings/@jupyterlab/docmanager-extension:plugin?1547931482527 (::1) 1.38ms
[D 21:58:02.548 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.549 LabApp] 200 GET /lab/api/settings/@jupyterlab/apputils-extension:themes?1547931482527 (::1) 1.25ms
[D 21:58:02.557 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.558 LabApp] 200 GET /api/sessions?1547931482556 (::1) 1.24ms
[D 21:58:02.609 LabApp] Path @jupyterlab/theme-light-extension/index.css served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/themes/@jupyterlab/theme-light-extension/index.css
[D 21:58:02.609 LabApp] 304 GET /lab/api/themes/@jupyterlab/theme-light-extension/index.css (::1) 0.73ms
[D 21:58:02.881 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.886 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.892 LabApp] 200 GET /api/contents/?content=1&1547931482882 (::1) 6.89ms
[D 21:58:02.893 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.895 LabApp] 200 GET /lab/api/settings/@jupyterlab/extensionmanager-extension:plugin?1547931482886 (::1) 2.17ms
[D 21:58:02.896 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:02.898 LabApp] 200 GET /lab/api/settings/@jupyterlab/notebook-extension:tracker?1547931482893 (::1) 1.95ms
[D 21:58:02.898 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.021 LabApp] 200 GET /api/nbconvert?1547931482894 (::1) 123.01ms
[D 21:58:03.022 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.023 LabApp] 200 GET /lab/api/settings/@jupyterlab/codemirror-extension:commands?1547931482918 (::1) 1.49ms
[D 21:58:03.024 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.024 LabApp] 200 GET /api/sessions?1547931482920 (::1) 0.64ms
[D 21:58:03.025 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.026 LabApp] 200 GET /lab/api/workspaces/lab?1547931482947 (::1) 1.22ms
[D 21:58:03.027 LabApp] Accepting token-authenticated connection from ::1
[I 21:58:03.038 LabApp] Node v10.14.2
    
[D 21:58:03.043 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.054 LabApp] Fetching URL: https://registry.npmjs.org/@jupyterlab%2Fplotly-extension
[D 21:58:03.131 LabApp] 200 GET /api/nbconvert?1547931482915 (::1) 88.10ms
[D 21:58:03.132 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.142 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.146 LabApp] 200 GET /api/contents/?1547931483062 (::1) 3.60ms
[D 21:58:03.146 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.147 LabApp] 200 GET /api/contents/plotly_minimsl_example.ipynb?content=0&1547931483063 (::1) 0.75ms
[D 21:58:03.147 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.149 LabApp] 200 GET /lab/api/settings/@jupyterlab/fileeditor-extension:plugin?1547931482896 (::1) 2.46ms
[I 21:58:03.150 LabApp] Build is up to date
[D 21:58:03.150 LabApp] 200 GET /lab/api/build?1547931482880 (::1) 269.08ms
[D 21:58:03.163 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.166 LabApp] 200 GET /api/contents/?content=1&1547931483162 (::1) 3.42ms
[D 21:58:03.250 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.462 LabApp] 200 GET /api/contents/plotly_minimsl_example.ipynb?type=notebook&content=1&1547931483248 (::1) 212.47ms
[D 21:58:03.905 LabApp] Path components/MathJax/MathJax.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/notebook/static/components/MathJax/MathJax.js
[D 21:58:03.906 LabApp] 304 GET /static/components/MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML-full,Safe&amp;delayStartupUntil=configured%22%20charset=%22utf-8%22 (::1) 0.97ms
[D 21:58:03.931 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.932 LabApp] 200 GET /api/contents/plotly_minimsl_example.ipynb/checkpoints?1547931483930 (::1) 1.24ms
[D 21:58:03.933 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:03.933 LabApp] 200 GET /api/contents/plotly_minimsl_example.ipynb/checkpoints?1547931483930 (::1) 0.71ms
[D 21:58:04.196 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:04.197 LabApp] 200 GET /api/contents/plotly_minimsl_example.ipynb/checkpoints?1547931484193 (::1) 1.96ms
[D 21:58:04.261 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:04.282 LabApp] Starting kernel: ['/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7', '-m', 'ipykernel_launcher', '-f', '/Users/andreas/Library/Jupyter/runtime/kernel-0a3237fa-497d-4e2c-b384-2671d1b0dd8c.json']
[D 21:58:04.287 LabApp] Connecting to: tcp://127.0.0.1:58482
[D 21:58:04.288 LabApp] Connecting to: tcp://127.0.0.1:58480
[I 21:58:04.290 LabApp] Kernel started: 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.290 LabApp] Kernel args: {'kernel_name': 'python3', 'cwd': '/Users/andreas/Teaching/02580/02580-pysrc'}
[D 21:58:04.291 LabApp] 201 POST /api/sessions?1547931484259 (::1) 30.50ms
[D 21:58:04.293 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:04.294 LabApp] 204 PUT /lab/api/workspaces/lab?1547931484276 (::1) 1.26ms
[D 21:58:04.294 LabApp] Path components/MathJax/config/TeX-AMS-MML_HTMLorMML-full.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/notebook/static/components/MathJax/config/TeX-AMS-MML_HTMLorMML-full.js
[D 21:58:04.295 LabApp] 304 GET /static/components/MathJax/config/TeX-AMS-MML_HTMLorMML-full.js?V=2.7.5 (::1) 0.94ms
[D 21:58:04.316 LabApp] Initializing websocket connection /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels
[D 21:58:04.317 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:04.317 LabApp] Requesting kernel info from 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.317 LabApp] Connecting to: tcp://127.0.0.1:58479
[D 21:58:04.318 LabApp] Initializing websocket connection /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels
[D 21:58:04.319 LabApp] Accepting token-authenticated connection from ::1
[D 21:58:04.320 LabApp] Waiting for pending kernel_info request
[D 21:58:04.340 LabApp] Path components/MathJax/config/Safe.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/notebook/static/components/MathJax/config/Safe.js
[D 21:58:04.341 LabApp] 304 GET /static/components/MathJax/config/Safe.js?V=2.7.5 (::1) 0.71ms
[D 21:58:04.419 LabApp] Path components/MathJax/extensions/Safe.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/notebook/static/components/MathJax/extensions/Safe.js
[D 21:58:04.420 LabApp] 304 GET /static/components/MathJax/extensions/Safe.js?V=2.7.5 (::1) 0.72ms
[D 21:58:04.437 LabApp] Path components/MathJax/jax/output/HTML-CSS/fonts/STIX-Web/fontdata.js served from /opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/notebook/static/components/MathJax/jax/output/HTML-CSS/fonts/STIX-Web/fontdata.js
[D 21:58:04.438 LabApp] 304 GET /static/components/MathJax/jax/output/HTML-CSS/fonts/STIX-Web/fontdata.js?V=2.7.5 (::1) 0.73ms
[D 21:58:04.656 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.657 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.657 LabApp] Received kernel info: {'status': 'ok', 'protocol_version': '5.1', 'implementation': 'ipython', 'implementation_version': '6.5.0', 'language_info': {'name': 'python', 'version': '3.7.2', 'mimetype': 'text/x-python', 'codemirror_mode': {'name': 'ipython', 'version': 3}, 'pygments_lexer': 'ipython3', 'nbconvert_exporter': 'python', 'file_extension': '.py'}, 'banner': "Python 3.7.2 (default, Dec 30 2018, 08:55:50) \nType 'copyright', 'credits' or 'license' for more information\nIPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.\n", 'help_links': [{'text': 'Python Reference', 'url': 'https://docs.python.org/3.7'}, {'text': 'IPython Reference', 'url': 'https://ipython.org/documentation.html'}, {'text': 'NumPy Reference', 'url': 'https://docs.scipy.org/doc/numpy/reference/'}, {'text': 'SciPy Reference', 'url': 'https://docs.scipy.org/doc/scipy/reference/'}, {'text': 'Matplotlib Reference', 'url': 'https://matplotlib.org/contents.html'}, {'text': 'SymPy Reference', 'url': 'http://docs.sympy.org/latest/index.html'}, {'text': 'pandas Reference', 'url': 'https://pandas.pydata.org/pandas-docs/stable/'}]}
[I 21:58:04.657 LabApp] Adapting to protocol v5.1 for kernel 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.658 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[I 21:58:04.658 LabApp] Adapting to protocol v5.1 for kernel 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.658 LabApp] 101 GET /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels?session_id=7f85291f-2d62-44fa-8c8c-249a6e991e87&token=5d6f40f4ba80888de5202ce50122285fc7a496c0a2050ed5 (::1) 342.45ms
[D 21:58:04.659 LabApp] Opening websocket /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels
[D 21:58:04.659 LabApp] Getting buffer for 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.659 LabApp] Connecting to: tcp://127.0.0.1:58479
[D 21:58:04.659 LabApp] Connecting to: tcp://127.0.0.1:58480
[D 21:58:04.659 LabApp] Connecting to: tcp://127.0.0.1:58481
[D 21:58:04.660 LabApp] 101 GET /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels?session_id=5bc0ca50-7077-4a95-b376-c982cb3bb67e&token=5d6f40f4ba80888de5202ce50122285fc7a496c0a2050ed5 (::1) 341.58ms
[D 21:58:04.660 LabApp] Opening websocket /api/kernels/0a3237fa-497d-4e2c-b384-2671d1b0dd8c/channels
[D 21:58:04.660 LabApp] Getting buffer for 0a3237fa-497d-4e2c-b384-2671d1b0dd8c
[D 21:58:04.660 LabApp] Connecting to: tcp://127.0.0.1:58479
[D 21:58:04.660 LabApp] Connecting to: tcp://127.0.0.1:58480
[D 21:58:04.660 LabApp] Connecting to: tcp://127.0.0.1:58481
[D 21:58:04.672 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.673 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.674 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.676 LabApp] activity on 0a3237fa-497d-4e2c-b384-2671d1b0dd8c: status
[D 21:58:04.684 LabApp] Websocket closed 0a3237fa-497d-4e2c-b384-2671d1b0dd8c:5bc0ca50-7077-4a95-b376-c982cb3bb67e
[D 21:58:05.050 LabApp] 200 GET /lab/api/extensions?1547931482955 (::1) 2023.00ms
</pre>
</details>

**Comment by jasongrout** [comment]
It's also very likely that this is a problem with the plotly extension, not with JLab itself (which I believe is at https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/plotly-extension ?). CC @gnestor.

**Comment by janba** [comment]
Absolutely, I lack a good enough overview of the software stack to tell which component is causing the issue, but probably "scene.container.parentNode.removeChild" is a dead give away.

**Comment by janba** [comment]
It works with FigureWidget. To be more precise, I can run the code below an arbitrary number of times and get correct graphical output every time. 
```
import plotly.graph_objs as go
mesh = go.Mesh3d(x=[0,1,0],y=[0,0,1],z=[0,0,0],i=[0],j=[1],k=[2],flatshading=False,color='#333333')
go.FigureWidget([mesh])
```
Since this is the more modern syntax (Plotly 3+) I am closing the issue even if there is probably a bug if we require backwards compatibility.

**Comment by gnestor** [comment]
Thanks @janba for reporting this! It looks like this is a legitimate bug. The fix is to replace `Plotly.newPlot` with `Plotly.react` in plotly-extension (thanks to https://github.com/stencila/stencila/issues/687). I have submitted a fix: https://github.com/jupyterlab/jupyter-renderers/pull/176

BTW, calling `py.init_notebook_mode` does not work in JupyterLab (it's designed for classic notebook) and it actually injects another version of plotly.py into the page which can cause some unexpected behavior.

**Comment by janba** [comment]
Excellent. Thank you! Both for the fix and the `init_notebook_mode` comment.

**Comment by gnestor** [comment]
Just published @jupyterlab/plotly-extension@0.18.2 to npm with this fix 👍

**Comment by janba** [comment]
Thanks! It works now and iplot keeps going after the maximum number of WebGL contexts have been used. When exporting to HTML it replaces the figure with an image placeholder.
